task/jefe/src/main.rs

```diff
@@ -110,72 +121,127 @@ fn main() -> ! {
+    disposition: &'s mut [Disposition; NUM_TASKS],
+    logged: &'s mut [bool; NUM_TASKS],
+    deadline: u64,
+}
 
-            // If our timer went off, we need to reestablish it
-            if msginfo.operation & TIMER_MASK != 0 {
-                deadline += TIMER_INTERVAL;
-                sys_set_timer(Some(deadline), TIMER_MASK);
+impl idl::InOrderJefeImpl for ServerImpl<'_> {
+    fn get_state(
+        &mut self,
+        _msg: &userlib::RecvMessage,
+    ) -> Result<u32, idol_runtime::RequestError<core::convert::Infallible>>
+    {
+        Ok(self.state)
+    }
+
+    fn set_state(
+        &mut self,
+        msg: &userlib::RecvMessage,
+        state: u32,
+    ) -> Result<(), idol_runtime::RequestError<core::convert::Infallible>> {
+        // If a task is designated as state owner, ensure that this RPC came
+        // from that task.
+        if let Some(owner) = generated::STATE_OWNER {
+            if msg.sender.index() != owner as usize {
+                // We'll pretend this operation doesn't exist.
+                // TODO this should be AccessViolation but that isn't exposed in
+                // the current version of idol.
+                return Err(idol_runtime::ClientError::UnknownOperation.fail());
             }
+        }
 
-            // If our disposition has changed or if we have been notified of
-            // a faulting task, we need to iterate over all of our tasks.
-            if changed || (msginfo.operation & fault_mask) != 0 {
-                for i in 0..hubris_num_tasks::NUM_TASKS {
-                    match kipc::read_task_status(i) {
-                        abi::TaskState::Faulted { fault, .. } => {
-                            if !logged[i] {
-                                log_fault(i, &fault);
-                                logged[i] = true;
-                            }
-
-                            if disposition[i] == Disposition::Restart {
-                                // Stand it back up
-                                kipc::restart_task(i, true);
-                                logged[i] = false;
-                            }
+        if self.state != state {
+            self.state = state;
+
+            for (task, mask) in generated::MAILING_LIST {
+                let taskid =
+                    TaskId::for_index_and_gen(task as usize, Generation::ZERO);
+                let taskid = sys_refresh_task_id(taskid);
+                sys_post(taskid, mask);
+            }
+        }
+        Ok(())
+    }
+}
+
+impl idol_runtime::NotificationHandler for ServerImpl<'_> {
+    fn current_notification_mask(&self) -> u32 {
+        FAULT_MASK | TIMER_MASK
+    }
+
+    fn handle_notification(&mut self, bits: u32) {
+        // Check to see if we have any external requests
+        let changed = external::check(self.disposition);
+
+        // If our timer went off, we need to reestablish it
+        if bits & TIMER_MASK != 0 {
+            self.deadline += TIMER_INTERVAL;
```

Comment:
`self.deadline` is initialized to `TIMER_INTERVAL` above.  If `jefe` restarts, I think this would cause it to spam timer interrupts until it catches up to kernel time.

We should either
- Initialize it to `sys_get_timer().now + TIMER_INTERVAL` when constructing the `ServerImpl`
- Update it with `self.deadline = sys_get_timer().now + TIMER_INTERVAL` when it goes off

Reply:
Well, technically, if `jefe` restarts, we're resetting the processor. However, initializing it to the current time would be more obviously correct, and I'll make that first change.

I'm not sure whether the second change is right -- this timeout thing is something @bcantrill added for the Hiffy interface, and adding the increment to it every time ensures that it doesn't drift. Adding the increment to the current time will cause drift if stuff is happening that delays execution of the code. In this case it likely doesn't matter, given that the timeout is set to be "fast for a human but slow for a computer," so I'm inclined to stick with the cheaper (fewer syscalls) option.